lib/freebl/algcmac.c

```diff
@@ -0,0 +1,327 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifdef FREEBL_NO_DEPEND
+#include "stubs.h"
+#endif
+
+#include "secport.h"
+#include "blapit.h"
+#include "blapi.h"
+#include "algcmac.h"
+#include "secerr.h"
+#include "nspr.h"
+
+#include "stdio.h"
+
+// Enum for identifying the underlying block cipher we're using internally.
+// To add a new cipher, add an entry here, define new CMAC_{name}Init and
+// CMAC_{name}Create methods, add a new Context pointer to the cipher union,
+// and update the cmac_Encrypt and CMAC_Destroy methods to handle the new
+// cipher.
+typedef enum {
+    CMAC_AES = 0
+} CMACCiphers;
+
+struct CMACContextStr {
+    // Information about the block cipher to use internally. The cipher should
+    // be placed in ECB mode so that we can use it to directly encrypt blocks.
+    // Store a pointer to the cipher's context and keep track of the block
+    // size so we can validate input and return parameters ourselves.
+    CMACCiphers cipherType;
+    union {
+        void *raw;
+        AESContext *aes;
+    } cipher;
+    int blockSize;
+
+    // Whether or not we allocated (via CMAC_Create) this CMACContext. This
+    // affects CMAC_Destroy.
+    PRBool wasAllocated;
+
+    // Internal keys which are conditionally used by the algorithm. Derived
+    // from encrypting the NULL block. We leave the storing of (and the
+    // cleanup of) the CMAC key to the underlying block cipher.
+    unsigned char *k1;
+    unsigned char *k2;
+
+    // When Update is called with data which isn't a multiple of the block
+    // size, we need a place to put it. HMAC handles this by passing it to
+    // the underlying hash function right away; we can't do that as the
+    // contract on the cipher object is different.
+    unsigned int partialIndex;
+    unsigned char *partialBlock;
+
+    // Last encrypted block. This gets xor-ed with partialBlock prior to
+    // encrypting it. NIST defines this to be the empty string to begin.
+    unsigned char *lastBlock;
+};
+
+/* byte arrays are really the wrong level for this: we want to shift the
+ * entire block left by one bit. With AES, this ends up being a 128-bit
+ * block size that we have to deal with manually; a union data structure
+ * between a __uint128_t and a byte array would be better. */
+static unsigned char *left_shift(unsigned char *data, int length) {
+    unsigned char *result = PORT_ZNewArray(unsigned char, length);
```

Comment:
So mallocs are expensive in the inner loops of code. You could make left_shift take the data it returns, then pass in data that is part either the stack or an element of a structure (so all the unsigned char *'s above would actually be unsigned char array[MAX_BLOCK_SIZE]).

Not a requirement, but you may get a similiar suggestion upstream.

Reply:
Does NSS have any internal helpers for this? This is really ugly code and is probably common to all of them.

If we delay the computation of `k1` and `k2` until the finish call, we can avoid the alloc by only computing only the one we need (avoiding needing to store both). This would make the CMAC computation not constant time though. 

We could also put `k1` and `k2` in the struct directly (`unsigned char k1[MAX_BLOCK_SIZE];`). This would make `CMACContext` larger when allocated on the stack, which I'm not sure if that is something we'd want.

Thoughts?

Reply:
Also, `left_shift` is only used twice, during initialization. It isn't called at all from the `CMAC_Update(...)` function.

Reply:
Yes, unsgined char k1[MAX_BLOCK_SIZE]; was what I was thinking. That reduced the malloc on Create from n to 1 (just the context).

On the helper question, were you asking about a helper for shift or a helper for dropping the mallocs?